src/postfork.cpp

```diff
@@ -63,39 +63,56 @@ static void debug_safe_int(int level, const char *format, int val) {
 /// process if it is the first in a JOB_CONTROL job.
 /// Returns true on sucess, false on failiure.
 bool child_set_group(job_t *j, process_t *p) {
-    bool retval = true;
     if (j->get_flag(job_flag_t::JOB_CONTROL)) {
         if (j->pgid == INVALID_PID) {
             j->pgid = p->pid;
         }
-        if (setpgid(p->pid, j->pgid) < 0) {
-            char pid_buff[128];
-            char job_id_buff[128];
-            char getpgid_buff[128];
-            char job_pgid_buff[128];
-            char argv0[64];
-            char command[64];
-
-            format_long_safe(pid_buff, p->pid);
-            format_long_safe(job_id_buff, j->job_id);
-            format_long_safe(getpgid_buff, getpgid(p->pid));
-            format_long_safe(job_pgid_buff, j->pgid);
-            narrow_string_safe(argv0, p->argv0());
-            narrow_string_safe(command, j->command_wcstr());
-
-            debug_safe(
-                1, "Could not send own process %s, '%s' in job %s, '%s' from group %s to group %s",
-                pid_buff, argv0, job_id_buff, command, getpgid_buff, job_pgid_buff);
-
-            safe_perror("setpgid");
-            retval = false;
+
+        while (int result = setpgid(p->pid, j->pgid) != 0) {
+            if (result == EPERM) {
+                // The setpgid(2) man page says that EPERM is returned only if attempts are made to
+                // move processes into groups across session boundaries (which can never be the case
+                // in fish, anywhere) or to change the process group ID of a session leader (again,
+                // can never be the case). I'm pretty sure this is a WSL bug, as we see the same
+                // with tcsetpgrp(2) in other places and it disappears on retry.
+                debug_safe(2, "setpgid(2) returned EPERM. Retrying");
+            }
+            else if (result == EINTR) {
```

Comment:
Don't we usually let the "else" ride on the closing bracket?

```c++
} else if
```

?

Reply:
I know we do with final `} else {` clauses, but I wasn't sure about `} else if (...) {` but a quick check with `ag '\}\n\s*else if' -c` reveals you are correct.